File: src/big_config/render.clj
(ns big-config.render
  (:require [big-config.core :as core]
            [big-config.keys :as k]
            [big-config.selmer-filters :as filters]
            [big-config.utils :as u]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [selmer.parser :as selmer]
            [selmer.util :as selmer-util])
  (:import [java.io File]
           [java.nio.file Files StandardCopyOption LinkOption]))

(def non-replaced-exts #{"jpg" "jpeg" "png" "gif" "bmp" "bin"})

(def template-keys
  #{:template "template"
    :target-dir "target-dir" :targetDir "targetDir"
    :overwrite "overwrite"
    :data-fn "data-fn" :dataFn "dataFn"
    :template-fn "template-fn" :templateFn "templateFn"
    :post-process-fn "post-process-fn" :postProcessFn "postProcessFn"
    :transform "transform"})

(def delimiter-keys
  #{:tag-open :tagOpen :tag-close :tagClose
    :filter-open :filterOpen :filter-close :filterClose
    :tag-second :tagSecond :short-comment-second :shortCommentSecond})

(defn- get-any [m & ks]
  (some (fn [key]
          (when (contains? m key)
            (get m key)))
        ks))

(defn- char-opt [x]
  (first (str x)))

(defn- selmer-opts [delimiters]
  (into {} (map (fn [[dk dv]] [dk (char-opt dv)]) (filters/normalize-delimiters delimiters))))

(defn render-string
  ([input data] (render-string input data nil))
  ([input data delimiters]
   (selmer-util/without-escaping
     (selmer/render (filters/whitespace-control input delimiters) data (selmer-opts delimiters)))))

(defn- file-ext [file]
  (let [n (.getName (io/file file))
        i (.lastIndexOf n ".")]
    (if (neg? i) "" (str/lower-case (subs n (inc i))))))

(defn- ensure-parent [file]
  (when-let [parent (.getParentFile (io/file file))]
    (.mkdirs parent)))

(defn- walk-files [dir]
  (let [root (io/file dir)]
    (when-not (.exists root)
      (throw (ex-info (str "Template directory not found: " dir) {:dir dir})))
    (letfn [(walk [^File f]
              (cond
                (.isDirectory f) (mapcat walk (sort-by #(.getName ^File %) (.listFiles f)))
                (.isFile f) [f]
                :else []))]
      (vec (walk root)))))

(defn- relativize [root file]
  (str (.relativize (.toPath (io/file root)) (.toPath (io/file file)))))

(defn- copy-file [src target]
  (ensure-parent target)
  (Files/copy (.toPath (io/file src))
              (.toPath (io/file target))
              (into-array StandardCopyOption [StandardCopyOption/REPLACE_EXISTING StandardCopyOption/COPY_ATTRIBUTES]))
  (let [src-file (io/file src)
        target-file (io/file target)]
    (.setExecutable target-file (.canExecute src-file) false)
    (.setReadable target-file (.canRead src-file) false)
    (.setWritable target-file (.canWrite src-file) true)))

(defn copy-dir [{:keys [src-dir target-dir data delimiters]}]
  (doseq [src-file (walk-files src-dir)]
    (let [rel (relativize src-dir src-file)
          target-file (io/file target-dir rel)
          replaceable (not (contains? non-replaced-exts (file-ext src-file)))]
      (ensure-parent target-file)
      (if (and data replaceable)
        (spit target-file (render-string (slurp src-file) data delimiters))
        (copy-file src-file target-file)))))

(defn- entries-of-files [files]
  (cond
    (nil? files) []
    (map? files) (seq files)
    :else files))

(defn- delimiter-map? [x]
  (and (map? x) (some #(contains? x %) delimiter-keys)))

(defn- transform-option? [x]
  (contains? #{:raw "raw" ":raw" :only "only" ":only"} x))

(defn- normalize-transform-option [x]
  (keyword (str/replace (name (if (keyword? x) x (keyword (str x)))) #"^:" "")))

(defn parse-transform [spec]
  (when (or (not (sequential? spec)) (empty? spec))
    (throw (ex-info "Invalid transform entry" {:spec spec})))
  (let [[src & rest0] spec
        opts (->> rest0 (filter transform-option?) (map normalize-transform-option) set)
        rest (vec (remove transform-option? rest0))
        [target rest] (if (string? (first rest)) [(first rest) (subvec rest 1)] [nil rest])
        [files rest] (cond
                       (delimiter-map? (first rest)) [nil rest]
                       (map? (first rest)) [(first rest) (subvec rest 1)]
                       :else [nil rest])
        [delimiters _rest] (if (delimiter-map? (first rest)) [(first rest) (subvec rest 1)] [nil rest])]
    {:src src :target target :files files :delimiters delimiters :opts opts}))

(defn copy-template-dir [{:keys [template-dir target-dir data src target files delimiters opts]}]
  (let [data (or data {})
        opts (set (map normalize-transform-option (or opts [])))
        raw? (contains? opts :raw)
        only? (contains? opts :only)
        target-suffix (if target (render-string target data delimiters) "")
        target-base (if (seq target-suffix) (io/file target-dir target-suffix) (io/file target-dir))
        file-entries (vec (entries-of-files files))]
    (if (fn? src)
      (do
        (when (empty? file-entries)
          (throw (ex-info "Files is required when src is a function" {})))
        (doseq [[from to] file-entries]
          (let [raw-content (str (src from data))
                content (if raw? raw-content (render-string raw-content data delimiters))
                target-file (io/file target-base (render-string (str to) data delimiters))]
            (ensure-parent target-file)
            (spit target-file content))))
      (let [src-rendered (render-string (str src) data delimiters)
            src-dir (io/file template-dir src-rendered)]
        (if (empty? file-entries)
          (copy-dir {:src-dir src-dir :target-dir target-base :data (when-not raw? data) :delimiters delimiters})
          (do
            (when-not only?
              (copy-dir {:src-dir src-dir :target-dir target-base :data (when-not raw? data) :delimiters delimiters}))
            (doseq [[from to] file-entries]
              (let [rendered-to (render-string (str to) data delimiters)
                    src-file (io/file src-dir (str from))
                    target-file (io/file target-base rendered-to)]
                (when-not only?
                  (.delete (io/file target-base (str from))))
                (ensure-parent target-file)
                (if (and (not raw?) (not (contains? non-replaced-exts (file-ext src-file))))
                  (spit target-file (render-string (slurp src-file) data delimiters))
                  (copy-file src-file target-file))))))))))

(defn- directory? [x]
  (let [f (io/file x)]
    (and (.exists f) (.isDirectory f))))

(defn- url->file [url]
  (when (= "file" (.getProtocol url))
    (io/file (.toURI url))))

(defn resolve-template-dir [template]
  (let [template (str template)
        template-file (io/file template)
        absolute? (.isAbsolute template-file)
        cwd (System/getProperty "user.dir")
        candidates (remove nil?
                           [(when absolute? template-file)
                            (when-not absolute? (io/file cwd template))
                            (when-not absolute? (io/file cwd "resources" template))
                            (when-not absolute? (some-> (io/resource template) url->file))
                            (when-not absolute? (some-> (io/resource (str "big-config/" template)) url->file))])]
    (if-let [found (some #(when (directory? %) (.getCanonicalPath (io/file %))) candidates)]
      found
      (throw (ex-info "Template resource not found" {:template template
                                                       :candidates (mapv str candidates)})))))

(defn- template-target-dir [edn]
  (get-any edn :target-dir :targetDir "target-dir" "targetDir"))

(defn- get-multi-option [value]
  (cond
    (nil? value) []
    (and (sequential? value) (not (fn? value))) value
    :else [value]))

(defn- delete-recursive [file]
  (let [f (io/file file)]
    (when (.exists f)
      (when (.isDirectory f)
        (doseq [child (.listFiles f)]
          (delete-recursive child)))
      (.delete f))))

(defn render [opts]
  (let [templates (get opts k/render-templates)]
    (when (nil? templates)
      (throw (IllegalArgumentException. ":big-config.render/templates should never be nil")))
    (doseq [input-edn templates]
      (let [data-fn (u/->fn (get-any input-edn :data-fn :dataFn "data-fn" "dataFn")
                            (fn [data _opts] data))
            template-fn (u/->fn (get-any input-edn :template-fn :templateFn "template-fn" "templateFn")
                                (fn [_data edn] edn))
            data-base (into {} (remove (fn [[tk _]] (contains? template-keys tk)) input-edn))
            data (data-fn (merge data-base
                                 {:module (get opts k/render-module)
                                  :profile (get opts k/render-profile)})
                          opts)
            edn (template-fn data input-edn)
            template (get-any edn :template "template")
            target-dir (template-target-dir edn)
            transform (get-any edn :transform "transform")]
        (when (or (not (seq (str template))) (not (seq (str target-dir))))
          (throw (ex-info "Invalid template" {:edn edn})))
        (when (nil? transform)
          (throw (IllegalArgumentException. ":transform not defined")))
        (when (empty? transform)
          (throw (IllegalArgumentException. ":transform is an empty list")))
        (let [template-dir (resolve-template-dir template)
              target-file (io/file target-dir)
              overwrite (get-any edn :overwrite "overwrite")]
          (when (.exists target-file)
            (cond
              (#{:delete "delete" ":delete"} overwrite) (delete-recursive target-file)
              (not overwrite) (throw (ex-info (str target-dir " already exists (and :overwrite was not true).") {}))))
          (doseq [spec transform]
            (let [{:keys [src target files delimiters opts]} (parse-transform spec)]
              (copy-template-dir {:template-dir template-dir
                                  :target-dir target-dir
                                  :data data
                                  :src src
                                  :target target
                                  :files files
                                  :delimiters delimiters
                                  :opts opts})))
          (doseq [f0 (get-multi-option (get-any edn :post-process-fn :postProcessFn "post-process-fn" "postProcessFn"))]
            (let [f (u/->fn f0 (fn [_edn _data] nil))]
              (f edn data))))))
    (assoc opts k/exit 0 k/err nil)))

(def templates
  (core/create-workflow
   {:first-step :big-config.render/start
    :wire-fn (fn [step]
               (case step
                 :big-config.render/start [render :big-config.render/end]
                 :big-config.render/end [identity nil]
                 [identity nil]))}))

(defn discover [parent-dir]
  (let [parent (io/file parent-dir)
        out (atom [])]
    (letfn [(walk [dir depth]
              (when (<= depth 2)
                (doseq [entry (sort-by #(.getName ^File %) (.listFiles (io/file dir)))]
                  (when (.isDirectory entry)
                    (swap! out conj (relativize parent entry))
                    (walk entry (inc depth))))))]
      (walk parent 1)
      (vec (remove str/blank? @out)))))

(def whitespace-control filters/whitespace-control)

File: src/big_config/render.py
from __future__ import annotations

import importlib
import importlib.resources
import os
import shutil
import sys
import tempfile
from collections.abc import Callable, Iterable, Mapping
from pathlib import Path
from typing import Any

from selmer import render as selmer_render
from selmer import without_escaping

from . import core
from .selmer_filters import whitespace_control  # registers filters as a side effect
from .utils import BigConfigError, to_fn

TEMPLATES = "big-config.render/templates"
MODULE = "big-config.render/module"
PROFILE = "big-config.render/profile"
STEP_MODULE = "big-config.step/module"
STEP_PROFILE = "big-config.step/profile"

NON_REPLACED_EXTS: set[str] = {"jpg", "jpeg", "png", "gif", "bmp", "bin"}

# Top-level package under which target packages ship their template data. It is
# force-included into the wheel as a namespace package, so consumers can resolve
# templates through importlib.resources instead of materialising a ./resources
# symlink in the working directory.
RESOURCES_PACKAGE = "resources"

_TEMPLATE_KEYS = {
    "template",
    "target-dir",
    "target_dir",
    "overwrite",
    "data-fn",
    "data_fn",
    "template-fn",
    "template_fn",
    "post-process-fn",
    "post_process_fn",
    "transform",
}
_DELIMITER_KEYS = {"tag-open", "tag-close", "filter-open", "filter-close", "tag-second", "short-comment-second", "tag_open", "tag_close", "filter_open", "filter_close", "tag_second", "short_comment_second"}


def _selmer_opts(delimiters: Mapping[str, Any] | None = None) -> dict[str, Any]:
    delimiters = delimiters or {}
    return {
        "tag_open": str(delimiters.get("tag-open", delimiters.get("tag_open", "{"))),
        "tag_close": str(delimiters.get("tag-close", delimiters.get("tag_close", "}"))),
        "filter_open": str(delimiters.get("filter-open", delimiters.get("filter_open", "{"))),
        "filter_close": str(delimiters.get("filter-close", delimiters.get("filter_close", "}"))),
        "tag_second": str(delimiters.get("tag-second", delimiters.get("tag_second", "%"))),
        "short_comment_second": str(delimiters.get("short-comment-second", delimiters.get("short_comment_second", "#"))),
    }


def selmer(s: str, data: Mapping[str, Any] | None = None, delimiters: Mapping[str, Any] | None = None) -> str:
    prepared = whitespace_control(s, dict(delimiters or {}))
    return without_escaping(lambda: selmer_render(prepared, dict(data or {}), **_selmer_opts(delimiters)))


def _extension(path: Path | str) -> str:
    suffix = Path(path).suffix
    return suffix[1:].lower() if suffix.startswith(".") else suffix.lower()


def copy_dir(*, src_dir: str | os.PathLike[str], target_dir: str | os.PathLike[str], data: Mapping[str, Any] | None = None, delimiters: Mapping[str, Any] | None = None) -> None:
    src = Path(src_dir)
    target = Path(target_dir)
    for path in src.rglob("*"):
        if not path.is_file():
            continue
        rel = path.relative_to(src)
        target_file = target / rel
        target_file.parent.mkdir(parents=True, exist_ok=True)
        replaceable = _extension(path) not in NON_REPLACED_EXTS
        if data is not None and replaceable:
            content = path.read_text(encoding="utf-8")
            target_file.write_text(selmer(content, data, delimiters), encoding="utf-8")
            shutil.copymode(path, target_file)
        elif data is None and replaceable:
            # Text-like copy preserving exact bytes and permissions.
            shutil.copy2(path, target_file)
        else:
            shutil.copy2(path, target_file)


def _is_delimiters(value: Any) -> bool:
    return isinstance(value, Mapping) and bool(set(value.keys()) & _DELIMITER_KEYS)


def _is_option(value: Any) -> bool:
    return str(value).lstrip(":") in {"raw", "only"}


def _parse_transform_spec(spec: Iterable[Any]) -> dict[str, Any]:
    items = list(spec)
    if not items:
        raise ValueError("transform item must not be empty")
    src = items.pop(0)
    target = None
    files = None
    delimiters = None
    opts: list[str] = []
    if items and isinstance(items[0], str) and not _is_option(items[0]):
        target = items.pop(0)
    if items and isinstance(items[0], Mapping) and not _is_delimiters(items[0]):
        files = dict(items.pop(0))
    if items and _is_delimiters(items[0]):
        delimiters = dict(items.pop(0))
    while items:
        item = items.pop(0)
        if not _is_option(item):
            raise ValueError(f"invalid transform option: {item}")
        opts.append(str(item).lstrip(":"))
    return {"src": src, "target": target, "files": files, "delimiters": delimiters, "opts": opts}


def _source_as_fn(src: Any, template_dir: Path) -> Callable[[Any, Mapping[str, Any]], str] | None:
    if callable(src):
        return src
    if not isinstance(src, str):
        return None
    # Treat strings as directories when they exist under the template directory.
    if (template_dir / src).exists():
        return None
    try:
        return to_fn(src)
    except Exception:  # noqa: BLE001
        return None


def _delete_path(path: Path) -> None:
    if path.is_dir() and not path.is_symlink():
        shutil.rmtree(path)
    elif path.exists() or path.is_symlink():
        path.unlink()


def copy_template_dir(*, template_dir: str | os.PathLike[str], target_dir: str | os.PathLike[str], data: Mapping[str, Any], src: Any, target: str | None = None, files: Mapping[Any, str] | None = None, delimiters: Mapping[str, Any] | None = None, opts: Iterable[str] | None = None) -> None:
    template_path = Path(template_dir)
    target_path = Path(target_dir)
    opts_set = {str(x).lstrip(":") for x in (opts or [])}
    raw = "raw" in opts_set
    only = "only" in opts_set
    rendered_target = f"/{selmer(target, data, delimiters)}" if target is not None else ""

    src_fn = _source_as_fn(src, template_path)
    if src_fn is not None:
        if not files:
            raise BigConfigError("Files is required when src is a symbol", {})
        for key, to in files.items():
            content = src_fn(key, data)
            if not raw:
                content = selmer(str(content), data, delimiters)
            target_file = target_path / rendered_target.lstrip("/") / selmer(str(to), data, delimiters)
            target_file.parent.mkdir(parents=True, exist_ok=True)
            target_file.write_text(str(content), encoding="utf-8")
        return

    rendered_src = selmer(str(src), data, delimiters)
    source_dir = template_path / rendered_src
    if files:
        with tempfile.TemporaryDirectory(prefix="big-config") as tmp:
            intermediate = Path(tmp)
            inter_target = intermediate / rendered_target.lstrip("/")
            if not only:
                copy_dir(src_dir=source_dir, target_dir=inter_target)
            for from_name, to_name in files.items():
                _delete_path(inter_target / str(from_name))
                destination = inter_target / selmer(str(to_name), data, delimiters)
                destination.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy2(source_dir / str(from_name), destination)
            if raw:
                copy_dir(src_dir=intermediate, target_dir=target_path)
            else:
                copy_dir(src_dir=intermediate, target_dir=target_path, data=data, delimiters=delimiters)
    else:
        destination = target_path / rendered_target.lstrip("/")
        if raw:
            copy_dir(src_dir=source_dir, target_dir=destination)
        else:
            copy_dir(src_dir=source_dir, target_dir=destination, data=data, delimiters=delimiters)


def _candidate_template_dirs(template: str) -> list[Path]:
    p = Path(template)
    candidates = []
    if p.is_absolute() or p.exists():
        candidates.append(p)
    cwd = Path.cwd()
    candidates.extend(
        [
            cwd / template,
            cwd / "src" / "resources" / template,
            cwd / "env" / "test" / "resources" / template,
            cwd / "test" / "resources" / template,
            cwd / "resources" / template,
        ]
    )
    seen: set[str] = set()
    unique: list[Path] = []
    for candidate in candidates:
        key = str(candidate)
        if key not in seen:
            unique.append(candidate)
            seen.add(key)
    return unique


def _resource_roots() -> list[Path]:
    # Filesystem roots of the installed ``resources`` package, if importable.
    # A regular on-disk install yields a single concrete path; namespace and
    # editable installs yield a MultiplexedPath spanning several directories
    # (e.g. site-packages plus an editable source tree), enumerated via the
    # package's __path__. Returns an empty list when ``resources`` is absent so
    # callers fall back to the working-directory candidates.
    try:
        anchor = importlib.resources.files(RESOURCES_PACKAGE)
    except (ImportError, TypeError, ValueError):
        return []
    try:
        return [Path(os.fspath(anchor))]
    except TypeError:
        package = sys.modules.get(RESOURCES_PACKAGE) or importlib.import_module(RESOURCES_PACKAGE)
        return [Path(entry) for entry in getattr(package, "__path__", [])]


def _resource_template_dir(template: str) -> Path | None:
    for root in _resource_roots():
        candidate = root / template
        if candidate.is_dir():
            return candidate.resolve()
    return None


def find_template_dir(template: str) -> Path:
    for candidate in _candidate_template_dirs(template):
        if candidate.exists() and candidate.is_dir():
            return candidate.resolve()
    resource_dir = _resource_template_dir(template)
    if resource_dir is not None:
        return resource_dir
    raise BigConfigError("Template resource not found", {"template": template})


def _multi_option(opts: Mapping[str, Any], *keys: str) -> list[Any]:
    value = None
    for key in keys:
        if key in opts:
            value = opts[key]
            break
    if value is None:
        return []
    if isinstance(value, (list, tuple)):
        return list(value)
    return [value]


def _template_get(edn: Mapping[str, Any], dashed: str, underscored: str | None = None, default: Any = None) -> Any:
    if dashed in edn:
        return edn[dashed]
    if underscored and underscored in edn:
        return edn[underscored]
    return default


def render(opts: Mapping[str, Any]) -> dict[str, Any]:
    templates0 = opts.get(TEMPLATES)
    if templates0 is None:
        raise ValueError(":big-config.render/templates should never be nil")
    for edn0 in templates0:
        edn = dict(edn0)
        data_fn = to_fn(_template_get(edn, "data-fn", "data_fn"), lambda data, _opts: data)
        template_fn = to_fn(_template_get(edn, "template-fn", "template_fn"), lambda _data, current_edn: current_edn)
        data = {k: v for k, v in edn.items() if k not in _TEMPLATE_KEYS}
        data.update(
            {
                "module": opts.get(STEP_MODULE, opts.get(MODULE)),
                "profile": opts.get(STEP_PROFILE, opts.get(PROFILE)),
            }
        )
        data = data_fn(data, opts)
        edn = dict(template_fn(data, edn))
        template = _template_get(edn, "template")
        target_dir = _template_get(edn, "target-dir", "target_dir")
        if not isinstance(template, str) or not template.strip() or not isinstance(target_dir, str) or not target_dir.strip():
            raise BigConfigError("Invalid template", {"template": template, "target-dir": target_dir})
        transform = _template_get(edn, "transform")
        if transform is None:
            raise ValueError(":transform not defined")
        transform = list(transform)
        if not transform:
            raise ValueError(":transform is an empty list")
        template_dir = find_template_dir(template)
        target_path = Path(target_dir)
        overwrite = _template_get(edn, "overwrite")
        if target_path.exists():
            if overwrite:
                if str(overwrite).lstrip(":") == "delete":
                    shutil.rmtree(target_path)
            else:
                raise BigConfigError(f"{target_dir} already exists (and :overwrite was not true).", {})
        for spec0 in transform:
            spec = _parse_transform_spec(spec0)
            copy_template_dir(template_dir=template_dir, target_dir=target_path, data=data, **spec)
        for post in _multi_option(edn, "post-process-fn", "post_process_fn"):
            post_fn = to_fn(post, lambda _edn, _data: None)
            post_fn(edn, data)
    return core.ok(opts)


templates = core.workflow(
    {
        "first_step": "big-config.render/start",
        "wire_fn": lambda step, _step_fns: (render, "big-config.render/end") if step == "big-config.render/start" else (lambda x: x, None),
    }
)


def discover(parent_dir: str | os.PathLike[str]) -> list[str]:
    parent = Path(parent_dir)
    result: list[str] = []
    for p in parent.rglob("*"):
        if p.is_dir():
            rel = str(p.relative_to(parent))
            if rel:
                result.append(rel)
    return result
